pkg/storage/chunk/prometheus_chunk.go

```diff
@@ -162,24 +218,57 @@ func (p *prometheusChunkIterator) Value() model.SamplePair {
 	}
 }
 
-// TODO native histograms support, assumes it's used on float chunk and only keeps floats
-// The Batch function takes a valueType since the first sample we process is already acquired
-// by a Next/Seek/etc and we cannot get its type otherwise.
+func (p *prometheusChunkIterator) AtHistogram() (int64, *histogram.Histogram) {
+	return p.it.AtHistogram()
+}
+
+func (p *prometheusChunkIterator) AtFloatHistogram() (int64, *histogram.FloatHistogram) {
+	return p.it.AtFloatHistogram()
+}
+
+func (p *prometheusChunkIterator) Timestamp() int64 {
+	return p.it.AtT()
+}
+
 func (p *prometheusChunkIterator) Batch(size int, valueType chunkenc.ValueType) Batch {
 	var batch Batch
-	j := 0
-	for j < size {
-		if valueType == chunkenc.ValFloat {
+	batch.ValueType = valueType
+	var populate func(j int)
+	switch valueType {
+	case chunkenc.ValFloat:
+		populate = func(j int) {
 			t, v := p.it.At()
 			batch.Timestamps[j] = t
 			batch.Values[j] = v
-			j++
 		}
+	case chunkenc.ValHistogram:
+		populate = func(j int) {
+			t, h := p.it.AtHistogram()
+			batch.Timestamps[j] = t
+			batch.PointerValues[j] = unsafe.Pointer(h)
+		}
+	case chunkenc.ValFloatHistogram:
+		populate = func(j int) {
+			t, fh := p.it.AtFloatHistogram()
+			batch.Timestamps[j] = t
+			batch.PointerValues[j] = unsafe.Pointer(fh)
+		}
+	default:
+		panic(fmt.Sprintf("invalid chunk encoding %v", valueType))
+	}
+
+	j := 0
+	for j < size {
+		populate(j)
+		j++
 		if j < size {
-			valueType = p.it.Next()
-			if valueType == chunkenc.ValNone {
+			vt := p.it.Next()
+			if vt == chunkenc.ValNone {
 				break
 			}
+			if vt != valueType {
+				panic(fmt.Sprintf("chunk encoding expected to be consistent in chunk start %v now %v", valueType, vt))
```

Comment:
Looking at Prometheus chunk iterator implementations: https://github.com/prometheus/prometheus/tree/main/tsdb/chunkenc . Those either return None or a hardcoded chunk specific value, so this should really never happen.